### PR TITLE
De-randomize doc builds

### PR DESCRIPTION
**Pull request by QuLogic**
Based on some issues I've found trying to get consistency out of #5743, this disables some needless randomness in the documentation (by setting seeds, or dropping it altogether). There are still a few things that remain:
 * The hash in the inheritance diagrams changes; I guess this is a "bug" in the Sphinx extension, but I have not really looked much deeper.
 * There are some dictionaries in attributes and/or default parameters that don't stay sorted. I would assume that's also deep in Sphinx and I can't do anything about it.
 * Something is weird with the size of the background of the labels in [`boxplot_demo2.py`](); I guess the font size and bbox calculation are not always done in the same order.
 * The location of some tick marks is ever so slightly different in 41 images (including thumbnails and duplicates, so not quite so many actual examples.) I don't know why this is, maybe rounding differences?

### CODE REVIEW

**Comment by jenshnielsen** [comment]
:+1: 

I think this fails due to the issue fixed by #5770

**Comment by mdboom** [comment]
Cool.  This is very welcome work.

> The hash in the inheritance diagrams changes; I guess this is a "bug" in the Sphinx extension, but I have not really looked much deeper.

Looking at the code, nothing obvious leaps out as to why this would be the case...  As an experiment, you could try monkey-patching the hash function in `sphinx.ext.inheritance_diagram.get_graph_hash` to forcibly sort or otherwise normalize the stuff that goes into the hash (and then make a PR against Sphinx if successful -- I'm not suggesting monkey patching it as a matter of course from matplotlib).

> There are some dictionaries in attributes and/or default parameters that don't stay sorted. I would assume that's also deep in Sphinx and I can't do anything about it.

If you can reduce this to a simple standalone example where the attributes don't always come out in order, I suspect the Sphinx folks would be amenable to fixing it.

> Something is weird with the size of the background of the labels in boxplot_demo2.py; I guess the font size and bbox calculation are not always done in the same order.

I can certainly reproduce this one.  I might spend a little time looking.

> The location of some tick marks is ever so slightly different in 41 images (including thumbnails and duplicates, so not quite so many actual examples.) I don't know why this is, maybe rounding differences?

Which ones?  I wonder if this is due to something not quite right in the recent limits changes or possibly something like #5767

**Comment by QuLogic** [comment]
Inheritance diagrams should be fixed by sphinx-doc/sphinx#2210.

**Comment by QuLogic** [comment]
> Which ones? I wonder if this is due to something not quite right in the recent limits changes or possibly something like #5767

I don't know how these few always come out different somehow, but
* build/html/_images/demo_axis_direction.png
* build/html/_images/demo_axisline_style.png
* build/html/_images/demo_curvelinear_grid1.png
* build/html/_images/demo_curvelinear_grid.png
* build/html/_images/demo_floating_axes1.png
* build/html/_images/demo_floating_axes.png
* build/html/_images/demo_floating_axis1.png
* build/html/_images/demo_floating_axis.png
* build/html/_images/parasite_simple21.png
* build/html/_images/parasite_simple2.png
* build/html/_images/simple_axis_direction01.png
* build/html/_images/whats_new_99_axes_grid.png
plus whatever other copies of these there are.

I was thinking more of #5647, though.

**Comment by QuLogic** [comment]
Looks like this is passing again with the rebase; I think it can go in if there are no other issues and we can leave the other stuff for other PRs.

File: examples/pylab_examples/boxplot_demo2.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 randomDists = ['Normal(1,1)', ' Lognormal(1,1)', 'Exp(1)', 'Gumbel(6,4)',
                'Triangular(2,9,11)']
 N = 500
+np.random.seed(0)
```
Comment (jenshnielsen) [inline examples/pylab_examples/boxplot_demo2.py]: I would suggest adding a comment that the random seeding is only to make the example reproducible

File: examples/pylab_examples/xcorr_demo.py [hunk 2]
```diff
@@ -1,6 +1,9 @@
 import matplotlib.pyplot as plt
 import numpy as np
 
+
+np.random.seed(0)
+
```
Comment (jenshnielsen) [inline examples/pylab_examples/xcorr_demo.py]: Same

File: lib/matplotlib/__init__.py [hunk 3]
```diff
@@ -1690,7 +1690,7 @@ def param(func):
                 arg_names = []
             elif len(_arg_names) > 1 and (positional_parameter_names is None):
                 # we got no manual parameter names but more than an 'ax' ...
-                if len(set(replace_names) - set(_arg_names[1:])) == 0:
+                if len(replace_names - set(_arg_names[1:])) == 0:
```
Comment (jenshnielsen) [inline lib/matplotlib/__init__.py]: Not really sure I understand why this is needed and if its safe if replace_names contains duplicates.

File: lib/matplotlib/artist.py [hunk 4]
```diff
@@ -1267,8 +1267,7 @@ def aliased_name(self, s):
 
         if s in self.aliasd:
             return s + ''.join([' or %s' % x
-                                for x
-                                in six.iterkeys(self.aliasd[s])])
+                                for x in sorted(self.aliasd[s])])
```
Comment (jenshnielsen) [inline lib/matplotlib/artist.py]: Does this affect performance?

**Comment by jenshnielsen** [comment]
A few inline comments otherwise :+1:  I agree that leaving the other stuff for a subsequent PR is fine. 

Looks like the examples that list above are mainly in the mpl_toolkits axes_grid1 and axisartist.
Any change to these may conflict with #5752 which moves these around a bit

File: lib/matplotlib/__init__.py (hunk 3 above)
Comment (QuLogic) [inline lib/matplotlib/__init__.py]: `replace_names` is already converted to a set [at the beginning of the function](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/__init__.py#L1623); it's not really needed in this PR though.

Comment (jenshnielsen) [inline lib/matplotlib/__init__.py]: Ok thats fine

File: examples/pylab_examples/boxplot_demo2.py (hunk 1 above)
Comment (QuLogic) [inline examples/pylab_examples/boxplot_demo2.py]: There are many of these in the examples without any comment; add them as well?

**Comment by tacaswell** [comment]
@QuLogic If you can address the few line comments we can get this merged.  #5752 needs a rebase already so lets get this one in so we can move forward on #5752 and #5743.

If you need to install sphinx from your repo to get that fix.

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (QuLogic) [inline lib/matplotlib/artist.py]: I'm not sure, but there's usually only a few aliases and a whole lot more string manipulation to create the full docstrings so I'd wager not.

Comment (tacaswell) [inline lib/matplotlib/artist.py]: I think these are only called building the docs and as a failure mode in `getp`.

Comment (tacaswell) [inline lib/matplotlib/artist.py]: and once at import time.

File: examples/pylab_examples/boxplot_demo2.py (hunk 1 above)
Comment (jenshnielsen) [inline examples/pylab_examples/boxplot_demo2.py]: There are many issues with the examples so I wouldn't expect to fix all of them but I would like to do this in new examples

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (jenshnielsen) [inline lib/matplotlib/artist.py]: Ok that seems fine

**Comment by mdboom** [comment]
> If you need to install sphinx from your repo to get that fix.

That's fine as long as we remember to rebase that against upstream Sphinx periodically.  While I'm normally against monkey-patching, in this case I think I'd prefer monkey patching those two small methods in Sphinx because it places less burden to keep up with Sphinx.

File: examples/pylab_examples/boxplot_demo2.py (hunk 1 above)
Comment (QuLogic) [inline examples/pylab_examples/boxplot_demo2.py]: Done.

**Comment by QuLogic** [comment]
> If you need to install sphinx from your repo to get that fix.

Nah, I don't think that's necessary; it only ensures that consecutive builds give consistent names, but it won't make too much difference to the results.

**Comment by jenshnielsen** [comment]
Thanks. We can always fix the examples later then if there are several others with the same issue

**Pull request merged**